draft-irtf-cfrg-opaque.md

```diff
@@ -1417,72 +1433,67 @@ Output:
-3. session_key = Derive-Secret(prk, "SessionKey", Hash(preamble))
-4. Km2 = Derive-Secret(handshake_secret, "ServerMAC", "")
-5. Km3 = Derive-Secret(handshake_secret, "ClientMAC", "")
-6. Output (Km2, Km3, session_key)
+2. h_preamble = Hash(Preamble)
+3. handshake_secret = Derive-Secret(prk, "HandshakeSecret", h_preamble)
+4. session_key = Derive-Secret(prk, "SessionKey", h_preamble)
+5. Km2 = Derive-Secret(handshake_secret, "ServerMAC", "")
+6. Km3 = Derive-Secret(handshake_secret, "ClientMAC", "")
+7. Output (Km2, Km3, session_key)
 ~~~
 
 ### 3DH Client Functions {#ake-client}
 
 ~~~
-Start(credential_request)
-
-Parameters:
-- Nn, the nonce length.
+Start()
 
 State:
-- state, a ClientState structure.
-
-Input:
-- credential_request, a CredentialRequest structure.
+- state, the AKE client state.
 
 Output:
-- ke1, a KE1 structure.
+- auth_init, a AuthInit structure.
 
 Steps:
 1. client_nonce = random(Nn)
 2. client_secret, client_keyshare = GenerateAuthKeyPair()
-3. Create KE1 ke1 with (credential_request, client_nonce, client_keyshare)
-4. Populate state with ClientState(client_secret, ke1)
-6. Output (ke1, client_secret)
+3. Create AuthInit auth_init with (client_nonce, client_keyshare)
+4. state.client_secret = client_secret
+5. Output auth_init
 ~~~
 
 ~~~
-CLientFinalize(client_identity, client_private_key, server_identity,
-               server_public_key, ke2)
+ClientFinalize(client_identity, client_private_key, server_identity,
+               server_public_key, ke1, ke2)
 
 State:
-- state, a ClientState structure.
+- state, the AKE client state.
 
 Input:
-- client_identity, the optional encoded client identity, which is
-  set to client_public_key if not specified.
-- client_private_key, the client's private key.
-- server_identity, the optional encoded server identity, which is
-  set to server_public_key if not specified.
-- server_public_key, the server's public key.
-- ke2, a KE2 message structure.
+- client_identity, the optional client identity.
+- client_private_key, the client's private key for the AKE protocol.
+- server_identity, the optional server identity.
+- server_public_key, the server's public key for the AKE protocol.
+- ke1, a KE1 structure.
+- ke2, a KE2 structure.
 
 Output:
-- ke3, a KE3 structure.
+- auth_finish, an AuthFinish structure.
 - session_key, the shared session secret.
 
 Exceptions:
-- HandshakeError, when the handshake fails
+- HandshakeError, when the handshake fails.
 
 Steps:
-1. ikm = TripleDHIKM(state.client_secret, ke2.server_keyshare,
-    state.client_secret, server_public_key, client_private_key, ke2.server_keyshare)
-2. preamble = Preamble(client_identity, state.ke1, server_identity, ke2.inner_ke2)
+1. ikm = TripleDHIKM(state.client_secret, ke2.auth_response.server_keyshare,
+                     state.client_secret, server_public_key,
+                     client_private_key, ke2.auth_response.server_keyshare)
+2. preamble = Preamble(client_identity, ke1, server_identity, ke2)
```

Comment:
ke1 was previously in the state. Why change that here?

Reply:
Because the `AKE client state` only carries values from the AKE.
The external function (the one that calls the AKE) is the one that knows how to build KE1 during Start().

Reply:
This seems like an aesthetic change. Can we please keep these separate?

Reply:
Not sure how to implement your suggestion.

Just to add more context and information about the fix:

There are two states, one wrapping the other: 
A: The Main State: (https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/515717e8b6f19d427d6190f99450a850cfb709aa/draft-irtf-cfrg-opaque.md?plain=1#L988)
which contains:
- password, the input password; and
- blind, the random blinding scalar returned by `Blind()`; and
- B: client_ake_state, the client's AKE state if necessary.
 
and B: the client_ake_state (https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/515717e8b6f19d427d6190f99450a850cfb709aa/draft-irtf-cfrg-opaque.md?plain=1#L1249)
containing:
- client_secret, an opaque byte string of length Nsk.

---

ke1 must live in the state, yes. the question is in which one?

We cannot store ke1 in state B, because ke1 is constructed as (https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/515717e8b6f19d427d6190f99450a850cfb709aa/draft-irtf-cfrg-opaque.md?plain=1#L570
)
```
- `KE1` is composed of the `CredentialRequest` and `AuthInit` messages.
```

AuthInit is created by the AKE, but CredentialRequest is not.
So the AKE cannot construct KE1, this is why KE1 must be passed as a parameter.

Hence, ke1 must live in the main state A. -- This part needed to be updated, and this is fixed now.

Reply:
Well, this PR changes state from "generic state" to "AKE client state," and I'm asking that we revert that change and address it separately.